src/data/map_data.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2016 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under (1) the MaidSafe.net Commercial License,
+// version 1.0 or later, or (2) The General Public License (GPL), version 3, depending on which
+// licence you accepted on initial access to the Software (the "Licences").
+//
+// By contributing code to the SAFE Network Software, or to this project generally, you agree to be
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.1.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use error::RoutingError;
+use rust_sodium::crypto::sign::{self, Signature};
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use super::DataIdentifier;
+use xor_name::XorName;
+
+pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+
+/// Mutable map (key-value store) data.
+#[derive(Clone, Debug, Eq, PartialEq, RustcEncodable, RustcDecodable)]
+pub struct MapData {
+    name: XorName,
+    data: HashMap<Key, Entry>,
+    permissions: HashMap<sign::PublicKey, Vec<Permission>>,
+    version: u64,
+    owner_keys: Vec<sign::PublicKey>,
+    signatures: Vec<Signature>,
+}
+
+impl MapData {
+    pub fn new(name: XorName) -> Self {
+        MapData {
+            name: name,
+            data: HashMap::new(),
+            permissions: HashMap::new(),
+            version: 0,
+            owner_keys: Vec::new(),
+            signatures: Vec::new(),
+        }
+    }
+
+    pub fn name(&self) -> &XorName {
+        &self.name
+    }
+
+    pub fn identifier(&self) -> DataIdentifier {
+        DataIdentifier::Map(self.name)
+    }
+
+    pub fn validate_permission(&self,
+                               sign_key: &sign::PublicKey,
+                               permission: Permission)
+                               -> bool {
+        self.permissions.get(sign_key)
+            .map(|ps| ps.contains(&permission))
+            .unwrap_or(false)
+    }
+
+    pub fn validate_size(&self) -> bool {
+        unimplemented!()
+    }
+
+    pub fn get(&self, _key: &Key) -> Result<&Value, RoutingError> {
+        unimplemented!()
+    }
+}
+
+impl PartialOrd for MapData {
+    fn partial_cmp(&self, other: &MapData) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+// TODO: ordering based on name only. Verify it is sufficient!
+impl Ord for MapData {
+    fn cmp(&self, other: &MapData) -> Ordering {
+        self.name.cmp(&other.name)
+    }
+}
+
+// TODO: hashing based on name only. Verify is is sufficient!
```

Comment:
I don't think it is: In the data manager, we'll probably exchange hashes in refresh messages to synchronise the state of a data chunk. For that, the `Hash` impl needs to take into account the data and owner fields.